### ISSUE TITLE
Disabling callback-polling

### ISSUE THREAD

**Issue by gautamg795** [faye/faye#431]
I'm trying to use Faye in React Native, where there's no `document` tag for callback-polling's JSON-P work. From what I understand, the Bayeux spec requires callback polling for a cross-origin connection, but is there any way to override this or work around it? (Simply using client.disable leaves me with a 'Could not find usable connection type' message)
Thanks.

**Comment by Bhullnatik** [comment]
We are having the same issue with my team. The solution we came to was to fork `faye` and force the use of `XmlHttpRequest`. We would be very interested in an out-the-box solution though. 😁

**Comment by jcoglan** [comment]
The `long-polling` transport is only considered usable if the endpoint is on the same origin as the document:

https://github.com/faye/faye/blob/e7e6c975a28e748770325c14e0ead09a3d747bb9/javascript/transport/xhr.js#L53-L55

`URI.isSameOrigin()` compares the supplied endpoint to the global `location` object:

https://github.com/faye/faye/blob/e7e6c975a28e748770325c14e0ead09a3d747bb9/javascript/util/uri.js#L6-L11

I'm not familiar with React Native -- is there a global `location` object, and if so what does it look like? Presumably code running in native apps is not a web page in a browser and therefore does not have any origin. It's possible that in such environments we should make `XHR.isUsable()` return `true`.

**Comment by gautamg795** [comment]
React Native does not have a window.location object, but I was able to actually get around my aforementioned issues with the following code, adapted from what I found in someone's gist:
```
var url = require('url') // via npm install --save url
var Faye = require('faye/browser/faye-browser'); // via npm install --save faye
var FAYE_URL = 'http://my_faye_url.com/';
window.location = url.parse(FAYE_URL);
window.location.port = ""; // must be set to empty string if not explicitly specified
var fayeClient = new Faye.Client(FAYE_URL);
fayeClient.disable('callback-polling'); // probably not necessary now that they're on same origin
var subscription = fayeClient.subscribe(...);
```
(It's important to note `url.parse()` sets `window.location.port` to `null` if not specified, whereas Faye will expect empty string in `URI.isSameOrigin()`; so if you are not specifying an explicit port, set it to `""`)


Obviously a native solution in Faye would be awesome—but I know you can't support every platform out there, and this does seem to work for now

**Comment by jcoglan** [comment]
It sounds as though making `URI.isSameOrigin()` return `true` if there's no `location` object would solve this issue.

**Comment by Bhullnatik** [comment]
I'm not sure to what extends this is avoidable but there may be a weird `location` object when using the debugger with React-native though:
<img width="357" alt="screen shot 2016-03-23 at 10 22 56 am" src="https://cloud.githubusercontent.com/assets/2040807/13980884/408a61e2-f0e1-11e5-915b-0d3387334c97.png">

**Comment by gautamg795** [comment]
Ah, didn't think about debugger. One React-Native specific solution would be to check if `window.navigator.product === "ReactNative"`, though I don't know if you're opposed to a product-specific check in `isUsable` / `isSameOrigin`

**Comment by jcoglan** [comment]
I would strongly prefer to avoid any form of vendor sniffing if it all possible, because it's brittle and tends to be hard to work around when it stops working. Is there any reliable way we can make this work using only feature detection?

**Comment by Bhullnatik** [comment]
I tried to look over React-native's [XHR implementation](https://github.com/facebook/react-native/blob/master/Libraries/Network/XMLHttpRequestBase.js) but I didn't see anything in particular to recognize the feature. Other than trying a request on a specific server to make sure the cross-domain is working, I don't see much more options. :/

**Comment by jcoglan** [comment]
I'm very wary of putting in vendor sniffing code for a platform we don't explicitly support and that's unlikely to get tested, and which may have even more unknown behaviour apart from adding a `location` object in debug mode.

You best bet for now is to patch the client like so:

```js
Faye.Transport.XHR.isUsable = function(dispatcher, endpoint, callback, context) {
  callback.call(context, true);
};
```

**Comment by jcoglan** [comment]
I've added a check for `navigator.product` in a0d3598.

This will be released in 1.2.0.
